File: backend/src/main/java/com/rayhan/erp/model/Article.java
package com.rayhan.erp.model;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;

@Entity
@Table(name = "articles")
@Getter
@Setter
@NoArgsConstructor
public class Article {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(unique = true, nullable = false, length = 30)
    private String reference;

    @Column(nullable = false, length = 150)
    private String designation;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 10)
    private TypeArticle type; // MP, PF, PSF

    @Column(length = 20)
    private String uniteMesure; // kg, unité, rouleau, m

    @Column(precision = 15, scale = 3)
    private BigDecimal prixUnitaire = BigDecimal.ZERO;

    @Column(precision = 15, scale = 3)
    private BigDecimal stockActuel = BigDecimal.ZERO;

    @Column(precision = 15, scale = 3)
    private BigDecimal stockMinimum = BigDecimal.ZERO;

    private boolean actif = true;

    @Column(length = 100)
    private String assetImage;

    public enum TypeArticle {
        MP,   // Matière Première (HDPE, LDPE, colorants)
        PSF,  // Produit Semi-Fini (film tubulaire)
        PF    // Produit Fini (sacs, film rétractable)
    }
}

File: frontend/lib/models/article.dart
class Article {
  final int? id;
  final String reference;
  final String designation;
  final String type; // MP, PSF, PF
  final String? uniteMesure;
  final double prixUnitaire;
  final double stockActuel;
  final double stockMinimum;
  final bool actif;
  final String? assetImage;

  Article({
    this.id,
    required this.reference,
    required this.designation,
    required this.type,
    this.uniteMesure,
    this.prixUnitaire = 0,
    this.stockActuel = 0,
    this.stockMinimum = 0,
    this.actif = true,
    this.assetImage,
  });

  factory Article.fromJson(Map<String, dynamic> json) => Article(
        id: json['id'],
        reference: json['reference'] ?? '',
        designation: json['designation'] ?? '',
        type: json['type'] ?? 'MP',
        uniteMesure: json['uniteMesure'],
        prixUnitaire: (json['prixUnitaire'] ?? 0).toDouble(),
        stockActuel: (json['stockActuel'] ?? 0).toDouble(),
        stockMinimum: (json['stockMinimum'] ?? 0).toDouble(),
        actif: json['actif'] ?? true,
        assetImage: json['assetImage'],
      );

  Map<String, dynamic> toJson() => {
        if (id != null) 'id': id,
        'reference': reference,
        'designation': designation,
        'type': type,
        'uniteMesure': uniteMesure,
        'prixUnitaire': prixUnitaire,
        'stockActuel': stockActuel,
        'stockMinimum': stockMinimum,
        'actif': actif,
        if (assetImage != null) 'assetImage': assetImage,
      };

  bool get enAlerte => stockActuel <= stockMinimum;

  static const Map<String, String> typeLabels = {
    'MP': 'Matière Première',
    'PSF': 'Produit Semi-Fini',
    'PF': 'Produit Fini',
  };

  String get typeLabel => typeLabels[type] ?? type;
}
